### PR TITLE
Remove dependency on boost::regex from Whiskers

### PR DESCRIPTION
**Pull request by kcy1019**
#### TL;DR

Thanks in advance for advices :)

- Whiskers regex - **as-is**, with `boost::regex`:
   ```javascript
   /<([a-zA-Z0-9_$-]+)>|<#([a-zA-Z0-9_$-]+)>(.*?)</\2>|<\?([a-zA-Z0-9_$-]+)>(.*?)(<!\4>(.*?))?</\4>/g
   ```
- Whiskers regex - **to-be**, with `std::regex`:
  ```javascript
  /<([a-zA-Z0-9_$-]+)>|<#([a-zA-Z0-9_$-]+)>((?:.|\r|\n)*?)</\2>|<\?([a-zA-Z0-9_$-]+)>((?:.|\n|\r)*?)(<!\4>((?:.|\n|\r)*?))?</\4>/g
  ```

#### Details

This PR replaces `boost::regex` with `std::regex`, as suggested in #6971 and #6997.
The biggest difference between `boost::regex` and `std::regex` is the behavior of dots:
- In `boost::regex`, `.*` matches over line  breaks
- In `std::regex`, `.*` does not match newline characters(`\r`, `\n`).

To address the issue, a change is made to the Whiskers regular expression:
1. Replace **`.`** with non-capturing group **`(?:.|\r|\n)`**

And two compilation options are added to the _MSVC_ case:

1. `-DBOOST_REGEX_NO_LIB `: It prevents _automatic_ linking to `libboost_regex`.
2. `-D_REGEX_MAX_STACK_COUNT=200000L`: Microsoft implementation of `std::regex` has too tight recursion depth limit. This option increases it to 200,000 - which, IMHO, is sufficient for every sane use case.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (*if possible*)
- [x] README / documentation was extended, *if necessary*
- [x] Changelog entry (*if change is visible to the user*)
- [x] Used meaningful commit messages

Part of #7259.

### CODE REVIEW

File: libdevcore/Whiskers.cpp [hunk 1]
```diff
@@ -106,51 +106,66 @@ string Whiskers::replace(
...
+	result.reserve(_template.length());
+	string::const_iterator lastMatchedPos(_template.cbegin());
+	while (curMatch != matchEnd)
 	{
-		string tagName(_match[1]);
-		string listName(_match[2]);
-		string conditionName(_match[4]);
-		if (!tagName.empty())
+		result.append(curMatch->prefix().first, curMatch->prefix().second);
+		result.append(([&](match_results<string::const_iterator> _match) -> string
```
Comment (chriseth) [inline libdevcore/Whiskers.cpp]: i think it might be less confusing to add an implementation of `boost::regex_replace` using `std::regex`. What do you think?

Comment (kcy1019) [reply libdevcore/Whiskers.cpp]: Yes, that would make logic easier to understand. I will try it shortly :)

File: libdevcore/Whiskers.cpp [hunk 2]
```diff
@@ -99,15 +99,43 @@ void Whiskers::checkParameterUnknown(string const& _parameter) const
 	);
 }
 
+namespace {
```
Comment (chriseth) [inline libdevcore/Whiskers.cpp]: Please fix formatting:
 - `{` on a line of its own
 - no indentation inside namespaces
 - `const` at the right side of the type
 - `)` on a line of its own
 - single indentation depth for function parameters

File: libdevcore/Whiskers.cpp [hunk 3]
```diff
@@ -99,15 +99,43 @@ void Whiskers::checkParameterUnknown(string const& _parameter) const
...
+	template<class Formatter>
+	string regex_replace(
+			const string& _s,
+			const regex& _e,
+			Formatter _fmt,
+			regex_constants::match_flag_type _flags = regex_constants::match_default)
+	{
+		string result;
+		result.reserve(_s.length());
```
Comment (chriseth) [inline libdevcore/Whiskers.cpp]: I think `.reserve` and `.shrink_to_fit` needlessly complicate the code.

File: libdevcore/Whiskers.cpp [hunk 4]
```diff
@@ -99,15 +99,43 @@ void Whiskers::checkParameterUnknown(string const& _parameter) const
 	);
 }
 
+namespace {
+	template<class Formatter>
+	string regex_replace(
+			const string& _s,
```
Comment (chriseth) [inline libdevcore/Whiskers.cpp]: Can you please use more descriptive variable names?

**Review by chriseth — APPROVED** [review]
Thanks a lot!
